File: ads/templates/renderer.py
"""
Template renderer - renders templates with variable substitution
"""

import re
from typing import Any, Dict, List


class TemplateRenderer:
    """Renders templates with variable substitution"""

    @staticmethod
    def render(template: str, variables: Dict[str, Any]) -> str:
        """
        Render template with variable substitution.

        Supports:
        - {{variable}} - simple substitution
        - {{variable|default_value}} - with default value
        - {{#if variable}}...{{/if}} - conditional blocks (future)

        Args:
            template: Template string
            variables: Variable values

        Returns:
            Rendered string
        """
        result = template

        # Handle {{variable|default}} pattern
        def replace_with_default(match):
            var_name = match.group(1)
            default = match.group(2) if match.lastindex >= 2 else ""
            return str(variables.get(var_name, default))

        result = re.sub(
            r'\{\{(\w+)\|([^}]*)\}\}',
            replace_with_default,
            result
        )

        # Handle simple {{variable}} pattern
        def replace_simple(match):
            var_name = match.group(1)
            return str(variables.get(var_name, f"{{{{{var_name}}}}}"))

        result = re.sub(
            r'\{\{(\w+)\}\}',
            replace_simple,
            result
        )

        return result

    @staticmethod
    def validate(template: str, variables: Dict[str, Any]) -> Dict[str, Any]:
        """
        Validate template rendering.

        Args:
            template: Template string
            variables: Variable values

        Returns:
            Validation result with 'valid', 'missing_variables', 'errors'
        """
        # Extract required variables (those without defaults)
        required = re.findall(r'\{\{(\w+)\}\}', template)

        # Extract variables with defaults
        with_defaults = re.findall(r'\{\{(\w+)\|[^}]*\}\}', template)

        # Remove variables that have defaults
        required = [v for v in required if v not in with_defaults]

        # Find missing variables
        missing = [v for v in required if v not in variables]

        return {
            "valid": len(missing) == 0,
            "missing_variables": missing,
            "required_variables": list(dict.fromkeys(required)),
            "optional_variables": list(dict.fromkeys(with_defaults)),
            "errors": [] if len(missing) == 0 else [
                f"Missing required variables: {', '.join(missing)}"
            ]
        }

    @staticmethod
    def extract_variables(template: str) -> List[str]:
        """
        Extract all variables from template.

        Args:
            template: Template string

        Returns:
            List of variable names in order of appearance
        """
        # Extract all variables
        all_vars = re.findall(r'\{\{(\w+)(?:\|[^}]*)?\}\}', template)
        return list(dict.fromkeys(all_vars))

File: src/templates/renderer.ts
export class TemplateRenderer {
  static render(template: string, variables: Record<string, unknown>): string {
    let result = template;

    result = result.replace(/\{\{(\w+)\|([^}]*)\}\}/g, (_match, varName: string, defaultValue: string) => {
      const value = variables[varName];
      return value !== undefined && value !== null ? String(value) : defaultValue;
    });

    result = result.replace(/\{\{(\w+)\}\}/g, (_match, varName: string) => {
      const value = variables[varName];
      return value !== undefined && value !== null ? String(value) : `{{${varName}}}`;
    });

    return result;
  }

  static validate(template: string, variables: Record<string, unknown>): Record<string, unknown> {
    const requiredMatches = template.match(/\{\{(\w+)\}\}/g) ?? [];
    const defaultsMatches = template.match(/\{\{(\w+)\|[^}]*\}\}/g) ?? [];

    const required = new Set<string>();
    const withDefaults = new Set<string>();

    for (const match of requiredMatches) {
      const name = match.replace(/[{}]/g, "");
      required.add(name);
    }

    for (const match of defaultsMatches) {
      const name = match.replace(/\{\{|\|.*|\}\}/g, "");
      withDefaults.add(name);
    }

    for (const name of withDefaults) {
      required.delete(name);
    }

    const missing: string[] = [];
    for (const name of required) {
      if (!(name in variables)) {
        missing.push(name);
      }
    }

    return {
      valid: missing.length === 0,
      missing_variables: missing,
      required_variables: Array.from(required),
      optional_variables: Array.from(withDefaults),
      errors: missing.length === 0 ? [] : [`Missing required variables: ${missing.join(", ")}`],
    };
  }

  static extractVariables(template: string): string[] {
    const matches = template.match(/\{\{(\w+)(?:\|[^}]*)?\}\}/g) ?? [];
    const unique: string[] = [];
    for (const match of matches) {
      const name = match.replace(/\{\{|\|.*|\}\}/g, "");
      if (!unique.includes(name)) {
        unique.push(name);
      }
    }
    return unique;
  }
}
